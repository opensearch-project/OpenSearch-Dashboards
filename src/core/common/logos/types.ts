/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { ColorScheme, ImageType } from './constants';

/**
 * @public
 */
export interface Logos {
  /**
   * @deprecated Use {@link Logos.Application} instead.
   */
  readonly OpenSearch: LogoItem;
  readonly Application: LogoItem;
  readonly Mark: LogoItem;
  /**
   * @deprecated Use {@link Logos.Mark} instead.
   */
  readonly CenterMark: LogoItem;
  readonly AnimatedMark: LogoItem;
  readonly colorScheme: ColorScheme;
}

export type LogoItem = ImageItem & Record<ColorScheme, ImageItem>;

export interface ImageItem {
  /**
   * The URL of the image
   */
  readonly url: string;
  readonly type: ImageType;
}
