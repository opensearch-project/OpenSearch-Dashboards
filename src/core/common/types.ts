/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export type { Logos } from './logos/types';
