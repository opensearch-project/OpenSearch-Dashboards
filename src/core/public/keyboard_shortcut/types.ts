/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Any modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

export interface KeyboardShortcutSetup {
  register(shortcut: ShortcutDefinition): void;
}

export interface KeyboardShortcutStart {
  register(shortcut: ShortcutDefinition): void;
  unregister(shortcut: Pick<ShortcutDefinition, 'id' | 'pluginId'>): void;
}

/**
 * A keyboard shortcut definition that plugins can register with the core system
 * @public
 */
export interface ShortcutDefinition {
  /**
   * Unique identifier for the shortcut within the plugin
   * Combined with pluginId to create a globally unique identifier
   */
  id: string;

  /**
   * The plugin that owns this shortcut
   * Used for namespacing and cleanup when plugins are disabled
   */
  pluginId: string;

  /**
   * Human-readable name for the shortcut
   * Used for documentation and help systems
   */
  name: string;

  /**
   * Category to group related shortcuts together
   * Examples: 'navigation', 'editing', 'view'
   */
  category: string;

  /**
   * Key combination that triggers the shortcut
   * Format: 'ctrl+s', 'shift+d', 'alt+enter', etc.
   * Case-insensitive, normalized internally
   */
  keys: string;

  /**
   * Function to execute when the shortcut is triggered
   * Can be async for operations that require API calls
   */
  execute: () => void | Promise<void>;
}
