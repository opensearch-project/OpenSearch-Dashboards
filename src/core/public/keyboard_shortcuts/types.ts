/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Any modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

// Core Type Definitions

export interface KeyboardShortcutsSetup {
  register(shortcuts: ShortcutDefinition[]): string[];
}

export interface KeyboardShortcutsStart {
  register(shortcuts: ShortcutDefinition[]): string[];
  unregister(id: string): void;
}

export interface ShortcutDefinition {
  id: string;
  pluginId: string;
  name: string;
  category: string;
  keys: string;
  execute: ShortcutHandler;
}

export type ShortcutHandler = (event: KeyboardEvent) => void | Promise<void>;
