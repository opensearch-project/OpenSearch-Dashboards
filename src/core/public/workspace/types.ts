/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { WorkspaceAttribute } from '../../types';

export type WorkspaceObject = WorkspaceAttribute & { readonly?: boolean; owner?: boolean };

export type IWorkspaceResponse<T> =
  | {
      result: T;
      success: true;
    }
  | {
      success: false;
      error?: string;
    };

export interface AssociationResult {
  id: string;
  error?: string;
}

/**
 * This interface representing a client for managing workspace-related operations.
 * Workspace client should implement this interface.
 *
 * TODO: Refactor the current workspace client implementation in workspace plugin to add the missing operations to this interface
 */
export interface IWorkspaceClient {
  /**
   * copy saved objects to target workspace
   *
   * @param {Array<{ id: string; type: string }>} objects
   * @param {string} targetWorkspace
   * @param {boolean} includeReferencesDeep
   * @returns {Promise<IResponse<any>>} result for this operation
   */
  copy(objects: any[], targetWorkspace: string, includeReferencesDeep?: boolean): Promise<any>;

  /**
   * Associates a list of objects with the given workspace ID.
   *
   * This method takes a workspace ID and an array of objects, where each object contains
   * an `id` and `type`. It attempts to associate each object with the specified workspace.
   * If the association succeeds, the object is included in the result without an error.
   * If there is an issue associating an object, an error message is returned for that object.
   *
   * @returns A promise that resolves to a response object containing an array of results for each object.
   *          Each result will include the object's `id` and, if there was an error during association, an `error` field
   *          with the error message.
   */
  associate(
    savedObjects: Array<{ id: string; type: string }>,
    workspaceId: string
  ): Promise<IWorkspaceResponse<AssociationResult[]>>;

  /**
   * Dissociates a list of objects from the given workspace ID.
   *
   * This method takes a workspace ID and an array of objects, where each object contains
   * an `id` and `type`. It attempts to dissociate each object from the specified workspace.
   * If the dissociation succeeds, the object is included in the result without an error.
   * If there is an issue dissociating an object, an error message is returned for that object.
   *
   * @returns A promise that resolves to a response object containing an array of results for each object.
   *          Each result will include the object's `id` and, if there was an error during dissociation, an `error` field
   *          with the error message.
   */
  dissociate(
    savedObjects: Array<{ id: string; type: string }>,
    workspaceId: string
  ): Promise<IWorkspaceResponse<AssociationResult[]>>;

  ui(): WorkspaceUI;
}

interface DataSourceAssociationProps {
  excludedDataSourceIds: string[];
  onComplete?: () => void;
  onError?: () => void;
}

export interface WorkspaceUI {
  DataSourceAssociation: (props: DataSourceAssociationProps) => JSX.Element;
}
