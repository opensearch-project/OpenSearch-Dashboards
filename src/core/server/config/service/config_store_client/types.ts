/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export interface ConfigDocument {
  config_name: string;
  config_blob: Record<string, any>;
}
