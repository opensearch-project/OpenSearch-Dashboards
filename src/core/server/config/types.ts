/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { ApiResponse } from '@opensearch-project/opensearch/.';
import { ConfigPath } from '@osd/config';
import { OpenSearchDashboardsRequest } from '../http';

/**
 * Setup allows core services and plugins to register config stores and any context to passed into async local store
 *
 * @interface
 */
export interface InternalDynamicConfigServiceSetup {
  registerDynamicConfigClientFactory: (factory: IDynamicConfigStoreClientFactory) => void;
  registerAsyncLocalStoreRequestHeader: (key: string | string[]) => void;
  getStartService: () => Promise<DynamicConfigServiceStart>;
}

export type DynamicConfigServiceSetup = InternalDynamicConfigServiceSetup;

export interface InternalDynamicConfigServiceStart {
  getClient: () => IDynamicConfigurationClient;
  getAsyncLocalStore: () => AsyncLocalStorageContext | undefined;
  createStoreFromRequest: (
    request: OpenSearchDashboardsRequest
  ) => AsyncLocalStorageContext | undefined;
}

export type DynamicConfigServiceStart = InternalDynamicConfigServiceStart;

export type IDynamicConfigurationClient = Pick<
  IInternalDynamicConfigurationClient,
  'getConfig' | 'bulkGetConfigs' | 'listConfigs'
>;

export interface IInternalDynamicConfigurationClient {
  createConfig: (
    createConfigProps: CreateConfigProps,
    options?: DynamicConfigurationClientOptions
  ) => Promise<ApiResponse>;
  bulkCreateConfigs: (
    bulkCreateConfigProps: BulkCreateConfigProps,
    options?: DynamicConfigurationClientOptions
  ) => Promise<ApiResponse>;
  getConfig: (
    getConfigProps: GetConfigProps,
    options?: DynamicConfigurationClientOptions
  ) => Promise<ConfigObject>;
  bulkGetConfigs: (
    bulkGetConfigProps: BulkGetConfigProps,
    options?: DynamicConfigurationClientOptions
  ) => Promise<Map<string, ConfigObject>>;
  listConfigs: (options?: DynamicConfigurationClientOptions) => Promise<Map<string, ConfigObject>>;
  deleteConfig: (
    deleteConfigs: DeleteConfigProps,
    options?: DynamicConfigurationClientOptions
  ) => Promise<ApiResponse>;
  bulkDeleteConfigs: (
    bulkDeleteConfigs: BulkDeleteConfigProps,
    options?: DynamicConfigurationClientOptions
  ) => Promise<ApiResponse>;
}

export interface IDynamicConfigStoreClientFactory {
  create: () => IDynamicConfigStoreClient;
}

/**
 * Client used to retrieve dynamic configs from the config store of choice
 *
 * @interface
 */
export type IDynamicConfigStoreClient = Pick<
  IInternalDynamicConfigurationClient,
  'listConfigs' | 'bulkCreateConfigs' | 'createConfig' | 'bulkDeleteConfigs' | 'deleteConfig'
> & {
  getConfig: (
    namespace: string,
    options?: DynamicConfigurationClientOptions
  ) => Promise<ConfigObject | undefined>;
  bulkGetConfigs: (
    namespaces: string[],
    options?: DynamicConfigurationClientOptions
  ) => Promise<Map<string, ConfigObject>>;
};

export interface DynamicConfigurationClientOptions {
  asyncLocalStorageContext: AsyncLocalStorageContext;
}

export type ConfigObject = Record<string, any>;

/**
 * Provides the necessary context needed when a request first hits the http server
 *
 * @interface AsyncLocalStorageContext
 */
export type AsyncLocalStorageContext = Map<string, any>;

export type ConfigIdentifier =
  | {
      /**
       * How plugin and core service schemas are identified.
       * - For plugins, this is the plugin id (the 'id' field in camelCase that can be found in the plugin manifest)
       *    - Use pluginConfigPath if a plugin has the 'configPath' property set in its manifest
       * - For core services, this is the path property found in the config schema
       *    - example: {@link HttpConfig} config name is 'server'
       */
      name: string;
      pluginConfigPath?: never;
    }
  | {
      /**
       * For plugins ONLY. This is the optional field 'configPath' in the plugin manifest. If a given plugin has 'configPath' in its manifest, set this value instead of name.
       */
      pluginConfigPath: ConfigPath;
      name?: never;
    };

export type ConfigBlob = ConfigIdentifier & {
  /**
   * The new config blob to override the old config in the config store. This will update the entire config blob for a configIdentifier
   */
  updatedConfig: ConfigObject;
};

export interface CreateConfigProps {
  config: ConfigBlob;
}

export interface BulkCreateConfigProps {
  configs: ConfigBlob[];
}

export type GetConfigProps = ConfigIdentifier;

export interface BulkGetConfigProps {
  paths: ConfigIdentifier[];
}

export type DeleteConfigProps = ConfigIdentifier;

export interface BulkDeleteConfigProps {
  paths: ConfigIdentifier[];
}
