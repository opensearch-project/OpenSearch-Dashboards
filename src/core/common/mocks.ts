/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export { getLogosMock } from './logos/get_logos.mock';
