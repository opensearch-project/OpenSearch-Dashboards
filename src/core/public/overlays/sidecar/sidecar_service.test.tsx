/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { mockReactDomRender, mockReactDomUnmount } from '../overlay.test.mocks';

import { mount } from 'enzyme';
import { i18nServiceMock } from '../../i18n/i18n_service.mock';
import {
  SidecarService,
  OverlaySidecarStart,
  OverlaySidecarOpenOptions,
  SIDECAR_DOCKED_MODE,
} from './sidecar_service';
import { OverlayRef } from '../types';

const i18nMock = i18nServiceMock.createStartContract();

beforeEach(() => {
  mockReactDomRender.mockClear();
  mockReactDomUnmount.mockClear();
});

const mountText = (text: string) => (container: HTMLElement) => {
  const content = document.createElement('span');
  content.textContent = text;
  container.append(content);
  return () => {};
};

const getServiceStart = () => {
  const service = new SidecarService();
  return service.start({ i18n: i18nMock, targetDomElement: document.createElement('div') });
};

describe('SidecarService', () => {
  let sidecar: OverlaySidecarStart;
  const options: OverlaySidecarOpenOptions = {
    config: {
      dockedMode: SIDECAR_DOCKED_MODE.RIGHT,
      paddingSize: 460,
    },
  };
  beforeEach(() => {
    sidecar = getServiceStart();
  });

  describe('open sidecar', () => {
    it('renders a sidecar to the DOM', () => {
      expect(mockReactDomRender).not.toHaveBeenCalled();
      sidecar.open(mountText('Sidecar content'), options);
      const content = mount(mockReactDomRender.mock.calls[0][0]);
      expect(content.html()).toMatchSnapshot();
    });

    it('does not unmount if targetDom is null', async () => {
      const sidecarRef = new SidecarService().start({ i18n: i18nMock, targetDomElement: null });
      const ref = sidecarRef.open(mountText('Sidecar content'), options);
      const content = mount(mockReactDomRender.mock.calls[0][0]);
      expect(content.html()).toMatchSnapshot();
      await ref.close();
      expect(mockReactDomUnmount).not.toHaveBeenCalled();
    });

    describe('with a currently active sidecar', () => {
      let ref1: OverlayRef;
      beforeEach(() => {
        ref1 = sidecar.open(mountText('Sidecar content'), options);
      });
      it('replaces the current sidecar with a new one', () => {
        sidecar.open(mountText('Sidecar content 2'), options);
        expect(mockReactDomUnmount).toHaveBeenCalledTimes(1);
        const modalContent = mount(mockReactDomRender.mock.calls[1][0]);
        expect(modalContent.html()).toMatchSnapshot();
        expect(() => ref1.close()).not.toThrowError();
        expect(mockReactDomUnmount).toHaveBeenCalledTimes(1);
      });
      it('resolves onClose on the previous ref', async () => {
        const onCloseComplete = jest.fn();
        ref1.onClose.then(onCloseComplete);
        sidecar.open(mountText('Sidecar content 2'), options);
        await ref1.onClose;
        expect(onCloseComplete).toBeCalledTimes(1);
      });
    });
  });
  describe('SidecarRef#close()', () => {
    it('resolves the onClose Promise', async () => {
      const ref = sidecar.open(mountText('Sidecar content'), options);

      const onCloseComplete = jest.fn();
      ref.onClose.then(onCloseComplete);
      await ref.close();
      await ref.close();
      expect(onCloseComplete).toHaveBeenCalledTimes(1);
    });
    it('can be called multiple times on the same SidecarRef', async () => {
      const ref = sidecar.open(mountText('Sidecar content'), options);
      expect(mockReactDomUnmount).not.toHaveBeenCalled();
      await ref.close();
      expect(mockReactDomUnmount.mock.calls).toMatchSnapshot();
      await ref.close();
      expect(mockReactDomUnmount).toHaveBeenCalledTimes(1);
    });
    it("on a stale Sidecar doesn't affect the active sidecar", async () => {
      const ref1 = sidecar.open(mountText('Sidecar content 1'), options);
      const ref2 = sidecar.open(mountText('Sidecar content 2'), options);
      const onCloseComplete = jest.fn();
      ref2.onClose.then(onCloseComplete);
      mockReactDomUnmount.mockClear();
      await ref1.close();
      expect(mockReactDomUnmount).toBeCalledTimes(0);
      expect(onCloseComplete).toBeCalledTimes(0);
    });
  });

  describe('SidecarRef#Hide()', () => {
    it('hide the sidecar when calling hide ', () => {
      expect(mockReactDomRender).not.toHaveBeenCalled();
      sidecar.open(mountText('Sidecar content'), options);
      sidecar.hide();
      const content = mount(mockReactDomRender.mock.calls[0][0]);
      expect(content.html()).toMatchSnapshot();
      expect(content.hasClass('osdSidecarFlyout--hide'));
    });
  });

  describe('SidecarRef#Show()', () => {
    it('recover sidecar when calling show after calling hide', () => {
      expect(mockReactDomRender).not.toHaveBeenCalled();
      sidecar.open(mountText('Sidecar content'), options);
      sidecar.hide();
      expect(mount(mockReactDomRender.mock.calls[0][0]).hasClass('osdSidecarFlyout--hide'));
      sidecar.show();
      const content = mount(mockReactDomRender.mock.calls[0][0]);
      expect(content.html()).toMatchSnapshot();
      expect(content.hasClass('osdSidecarFlyout--hide')).toEqual(false);
    });
  });
});
