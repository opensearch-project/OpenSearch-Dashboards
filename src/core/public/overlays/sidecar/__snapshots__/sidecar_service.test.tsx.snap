// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidecarService SidecarRef#Hide() hide the sidecar when calling hide  1`] = `"<div class=\\"osdSidecarFlyout osdSidecarFlyout--dockedRight osdSidecarFlyout--hide\\" style=\\"width: 460px;\\"><button class=\\"sidecar-resizableButton resizableButton--horizontal\\" data-test-subj=\\"resizableButton\\" type=\\"button\\"></button><div class=\\"osdSidecarMountWrapper\\"><span>Sidecar content</span></div></div>"`;

exports[`SidecarService SidecarRef#Show() recover sidecar when calling show after calling hide 1`] = `"<div class=\\"osdSidecarFlyout osdSidecarFlyout--dockedRight\\" style=\\"width: 460px;\\"><button class=\\"sidecar-resizableButton resizableButton--horizontal\\" data-test-subj=\\"resizableButton\\" type=\\"button\\"></button><div class=\\"osdSidecarMountWrapper\\"><span>Sidecar content</span></div></div>"`;

exports[`SidecarService SidecarRef#close() can be called multiple times on the same SidecarRef 1`] = `
Array [
  Array [
    <div />,
  ],
]
`;

exports[`SidecarService open sidecar does not unmount if targetDom is null 1`] = `"<div class=\\"osdSidecarFlyout osdSidecarFlyout--dockedRight\\" style=\\"width: 460px;\\"><button class=\\"sidecar-resizableButton resizableButton--horizontal\\" data-test-subj=\\"resizableButton\\" type=\\"button\\"></button><div class=\\"osdSidecarMountWrapper\\"><span>Sidecar content</span></div></div>"`;

exports[`SidecarService open sidecar renders a sidecar to the DOM 1`] = `"<div class=\\"osdSidecarFlyout osdSidecarFlyout--dockedRight\\" style=\\"width: 460px;\\"><button class=\\"sidecar-resizableButton resizableButton--horizontal\\" data-test-subj=\\"resizableButton\\" type=\\"button\\"></button><div class=\\"osdSidecarMountWrapper\\"><span>Sidecar content</span></div></div>"`;

exports[`SidecarService open sidecar with a currently active sidecar replaces the current sidecar with a new one 1`] = `"<div class=\\"osdSidecarFlyout osdSidecarFlyout--dockedRight\\" style=\\"width: 460px;\\"><button class=\\"sidecar-resizableButton resizableButton--horizontal\\" data-test-subj=\\"resizableButton\\" type=\\"button\\"></button><div class=\\"osdSidecarMountWrapper\\"><span>Sidecar content 2</span></div></div>"`;
